File: src/solutions/java/N242ValidAnagram.java
package solutions.java;

import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.util.function.Function.identity;
import static java.util.stream.Collectors.counting;

public class N242ValidAnagram {

    public static boolean isAnagram(String s, String t) {
        Function<String, Map<Integer, Long>> fun =
                a -> a.chars().boxed().collect(Collectors.groupingBy(identity(), counting()));
        return fun.apply(s).equals(fun.apply(t));
    }

    public static void main(String[] args) {
        assert isAnagram("anagram", "nagaram");
        assert !isAnagram("rat", "car");
    }
}

File: src/solutions/scala/N242ValidAnagram.scala
package solutions.scala

object N242ValidAnagram {

  def isAnagram(s: String, t: String): Boolean = {
    val fun = (a: String) => a.groupBy(identity).mapValues(_.length)
    fun(s) == fun(t)
  }

  def isAnagram2(s: String, t: String): Boolean = s.sorted.equals(t.sorted)

  def main(args: Array[String]): Unit = {
    assert(isAnagram2("anagram", "nagaram"))
    assert(!isAnagram("rat", "car"))
    assert(isAnagram("anagram", "nagaram"))
    assert(!isAnagram2("rat", "car"))
  }
}
